File: algorithm/sort/MergeSort.java
package algorithm.sort;

/**
 * MergeSort
 * 
 * If not left >= right
 * 1. Find the middle point to divide the array into two halves:  
 *         middle m = (l+r)/2
 * 2. Call mergeSort for first half:   
 *         Call sort(arr, l, m)
 * 3. Call mergeSort for second half:
 *         Call sort(arr, m+1, r)
 * 4. Merge the two halves sorted in step 2 and 3:
 *         Call merge(arr, l, m, r)
 *
 * O(NlogN)
 */
public class MergeSort {
    public static void main(String[] args) {
        int[] test = {84, 12, 442, 11, 1, 23};
        mergeSort(test);

        for (int e : test)                  // prints 1 11 12 23 84 442
            System.out.print(e + " ");

        System.out.println();
    }

    public static void mergeSort(int[] arr) {
        sort(arr, 0, arr.length-1);
    }

    private static void sort(int[] arr, int left, int right) {
        // base case
        if (left >= right)              // left == right는 쪼개다가 엘리먼트가 하나인 상황
            return;

        // 엘러먼트가 두개 이상이면 분할 및 합병
        int mid = (left + right) / 2;
        sort(arr, left, mid);
        sort(arr, mid+1, right);
        merge(arr, left, mid, right);
    }

    private static void merge(int[] arr, int left, int mid, int right) {
        int i = left;               // 왼쪽 배열의 시작위치
        int j = mid + 1;            // 오른쪽 배열의 시작위치

        int[] tmp = new int[arr.length];
        int k = left;

        while (i <= mid && j <= right) {
            if (arr[i] <= arr[j]) {
                tmp[k++] = arr[i++];
            } else {
                tmp[k++] = arr[j++];
            }
        }
        while (i <= mid) {
            tmp[k++] = arr[i++];
        }
        while (j <= right) {
            tmp[k++] = arr[j++];
        }

        for (int idx=left; idx<=right; idx++) {
            arr[idx] = tmp[idx];
        }
    }
}

File: python/algorithm/sort/merge_sort.py
import unittest


def merge_sort(lst: list) -> None:
    __sort(lst, 0, len(lst) - 1)


def __sort(lst: list, left: int, right: int) -> None:
    # base case
    if left >= right:
        return

    mid = (left + right) >> 1
    __sort(lst, left, mid)
    __sort(lst, mid + 1, right)
    merge(lst, left, mid, right)


def merge(lst: list, left: int, mid: int, right: int) -> None:
    i = left
    j = mid + 1

    tmp = [None] * len(lst)
    k = left

    while i <= mid and j <= right:
        if lst[i] <= lst[j]:
            tmp.insert(k, lst[i])
            # tmp[k] = lst[i]
            i += 1
        else:
            tmp.insert(k, lst[j])
            # tmp[k] = lst[j]
            j += 1
        k += 1

    while i <= mid:
        tmp.insert(k, lst[i])
        # tmp[k] = lst[i]
        k += 1
        i += 1

    while j <= right:
        tmp.insert(k, lst[j])
        # tmp[k] = lst[j]
        k += 1
        j += 1

    for idx in range(left, right + 1):
        lst[idx] = tmp[idx]


class MergeSortTest(unittest.TestCase):
    def setUp(self):
        self.lst = [4, 129, 23, 1010, 10, 1, 48, 22]

    def test_empty_list(self):
        empty_list = []

        merge_sort(empty_list)
        self.assertListEqual(empty_list, [])

    def test_merge_sort(self):
        with self.assertRaises(NameError):
            __sort(self.lst, 0, len(self.lst) - 1)

        merge_sort(self.lst)
        self.assertListEqual(self.lst, [1, 4, 10, 22, 23, 48, 129, 1010])
